src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -780,6 +790,18 @@ internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol
 
             if (collectionTypeKind == CollectionExpressionTypeKind.CollectionInitializer)
             {
+                if (this.TryBindIEnumerableCollectionInstance(targetType, node.Syntax, diagnostics, out var collectionCreation))
+                {
+                    if (collectionCreation.HasAnyErrors)
+                    {
+                        reportedErrors = true;
+                    }
+                }
+                else
+                {
+                    throw ExceptionUtilities.Unreachable();
+                }
```

Comment:
For new code/feature, I think an exception is okay. FWIW, before I'd modified this method at all as part of this PR, there was a crash later in the pipeline.

Reply:
> there was a crash later in the pipeline (unspecified compilation error because we'd have an error in the bound tree but not reported a diagnostic)

Would that have been reported as an diagnostic though rather than crashing the compiler? It seems like a diagnostic is preferable for those unexpected cases. 

Also, we're asserting elsewhere in this method that errors are reported, so asserting would be consistent.